cluster/src/shard_set.rs

```diff
@@ -0,0 +1,233 @@
+// Copyright 2022-2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{collections::HashMap, sync::Arc};
+
+use meta_client::types::{ShardId, ShardInfo, TableInfo, TablesOfShard};
+use snafu::{ensure, OptionExt};
+
+use crate::{Result, ShardVersionMismatch, TableAlreadyExists, TableNotFound, UpdateFrozenShard};
+
+/// [ShardTablesCache] caches the information about tables and shards, and the
+/// relationship between them is: one shard -> multiple tables.
+#[derive(Debug, Default, Clone)]
+pub struct ShardSet {
+    inner: Arc<std::sync::RwLock<ShardSetInner>>,
+}
+
+impl ShardSet {
+    // Fetch all the shard infos.
+    pub fn all_shards(&self) -> Vec<ShardRef> {
+        self.inner.read().unwrap().all_shard_infos()
+    }
+
+    // Get the shard by its id.
+    pub fn get(&self, shard_id: ShardId) -> Option<ShardRef> {
+        self.inner.read().unwrap().get(shard_id)
+    }
+
+    /// Remove the shard.
+    pub fn remove(&self, shard_id: ShardId) -> Option<ShardRef> {
+        self.inner.write().unwrap().remove(shard_id)
+    }
+
+    /// Insert the tables of one shard.
+    pub fn insert(&self, shard_id: ShardId, shard: ShardRef) {
+        self.inner.write().unwrap().insert(shard_id, shard)
+    }
+}
+
+#[derive(Debug, Default)]
+struct ShardSetInner {
+    // Tables organized by shard.
+    // TODO: The shard roles should be also taken into considerations.
+    shards: HashMap<ShardId, ShardRef>,
+}
+
+impl ShardSetInner {
+    fn all_shard_infos(&self) -> Vec<ShardRef> {
+        self.shards.values().cloned().collect()
+    }
+
+    fn get(&self, shard_id: ShardId) -> Option<ShardRef> {
+        self.shards.get(&shard_id).cloned()
+    }
+
+    fn remove(&mut self, shard_id: ShardId) -> Option<ShardRef> {
+        self.shards.remove(&shard_id)
+    }
+
+    fn insert(&mut self, shard_id: ShardId, shard: ShardRef) {
+        self.shards.insert(shard_id, shard);
+    }
+}
+
+#[derive(Debug)]
+pub struct Shard {
+    pub data: ShardData,
+    pub serializing_lock: tokio::sync::Mutex<()>,
```

Comment:
How about this name?
```suggestion
    pub serial_lock: tokio::sync::Mutex<()>,
```